### PR TITLE
Make limiter code more robust for all different values of rF

### PR DESCRIPTION
**Pull request by lindsayad**
Per @grmnptr, QUICK is not TVD after rF=5 and SOU fails beyond rF=2.
Consequently, a coefficient of 1e6 is pretty ludicrous. Cutting to
1e3 seems to make nonlinear convergence quite a bit better for
a lid problem with symmetry in x-velocity about the vertical center-line.

Refs advection sheme test case introduced in https://github.com/idaholab/moose/pull/20504

Changing to this implementation yields this convergence with mu=10 for SOU:
```
 0 Nonlinear |R| = 6.324555e+01
      SVD: condition number 2.762043186623e+05, 0 of 301 singular values are (nearly) zero
      SVD: smallest singular values: 2.896406784572e-04 2.896406784628e-04 2.909781482793e-04 3.643022162686e-04 4.873155397084e-04
      SVD: largest singular values : 7.902113478054e+01 7.902114112018e+01 7.902118416085e+01 8.000000625016e+01 8.000000625016e+01
      0 Linear |R| = 6.324555e+01
      1 Linear |R| = 1.005915e-11
 1 Nonlinear |R| = 3.800465e-02
      SVD: condition number 6.917562077325e+05, 0 of 301 singular values are (nearly) zero
      SVD: smallest singular values: 2.897803510275e-04 2.910660283552e-04 2.951239830845e-04 3.722649345432e-04 4.104117922424e-04
      SVD: largest singular values : 8.008239691994e+01 8.662661721132e+01 9.503738574867e+01 1.047961424402e+02 2.004573567022e+02
      0 Linear |R| = 3.800465e-02
      1 Linear |R| = 1.170032e-14
 2 Nonlinear |R| = 3.339679e-02
      SVD: condition number 9.525230044026e+06, 0 of 301 singular values are (nearly) zero
      SVD: smallest singular values: 1.201811892056e-05 1.497390078924e-04 2.943263737448e-04 2.988638538149e-04 3.075805210561e-04
      SVD: largest singular values : 7.909166791186e+01 7.909811016071e+01 7.990900760805e+01 8.008244978950e+01 1.144753474148e+02
      0 Linear |R| = 3.339679e-02
      1 Linear |R| = 1.113415e-14
 3 Nonlinear |R| = 6.115405e-04
      SVD: condition number 3.315858719431e+06, 0 of 301 singular values are (nearly) zero
      SVD: smallest singular values: 2.680027429545e-05 2.892337389280e-04 2.917310469395e-04 2.947842279218e-04 3.665110592070e-04
      SVD: largest singular values : 7.909178651885e+01 7.909803438075e+01 7.977618060077e+01 8.008237241761e+01 8.886592320571e+01
      0 Linear |R| = 6.115405e-04
      1 Linear |R| = 1.069933e-14
 4 Nonlinear |R| = 9.791690e-05
      SVD: condition number 1.372728712597e+06, 0 of 301 singular values are (nearly) zero
      SVD: smallest singular values: 6.474382356713e-05 2.893307612330e-04 2.922615344323e-04 2.948628347832e-04 3.664199452387e-04
      SVD: largest singular values : 7.909180919454e+01 7.909805304572e+01 7.991346448455e+01 8.008241937073e+01 8.887570557389e+01
      0 Linear |R| = 9.791690e-05
      1 Linear |R| = 2.510106e-15
 5 Nonlinear |R| = 8.616211e-07
      SVD: condition number 1.227093949694e+06, 0 of 301 singular values are (nearly) zero
      SVD: smallest singular values: 7.242984064256e-05 2.893458357513e-04 2.923330976825e-04 2.949003243778e-04 3.664409467089e-04
      SVD: largest singular values : 7.909181850366e+01 7.909805922198e+01 7.992902619339e+01 8.008243125807e+01 8.887821922981e+01
      0 Linear |R| = 8.616211e-07
      1 Linear |R| = 2.159545e-17
 6 Nonlinear |R| = 5.719615e-11
```
and this convergence with mu=10 for QUICK:
```
 0 Nonlinear |R| = 6.324555e+01
      SVD: condition number 2.762043186623e+05, 0 of 301 singular values are (nearly) zero
      SVD: smallest singular values: 2.896406784572e-04 2.896406784628e-04 2.909781482793e-04 3.643022162686e-04 4.873155397084e-04
      SVD: largest singular values : 7.902113478054e+01 7.902114112018e+01 7.902118416085e+01 8.000000625016e+01 8.000000625016e+01
      0 Linear |R| = 6.324555e+01
      1 Linear |R| = 1.005915e-11
 1 Nonlinear |R| = 3.268702e-02
      SVD: condition number 3.421550899307e+05, 0 of 301 singular values are (nearly) zero
      SVD: smallest singular values: 2.893660493512e-04 2.896452053778e-04 2.909843882365e-04 3.639469856468e-04 4.869394766647e-04
      SVD: largest singular values : 7.904092024951e+01 7.922745373226e+01 8.002019575585e+01 8.246862163626e+01 9.900806663865e+01
      0 Linear |R| = 3.268702e-02
      1 Linear |R| = 9.121454e-15
 2 Nonlinear |R| = 1.725109e-03
      SVD: condition number 5.906217251849e+05, 0 of 301 singular values are (nearly) zero
      SVD: smallest singular values: 1.356753662876e-04 2.868337705011e-04 2.939401091540e-04 2.960594441266e-04 3.656289502239e-04
      SVD: largest singular values : 7.903902987392e+01 7.904092326967e+01 7.916407133404e+01 8.002019589088e+01 8.013281890187e+01
      0 Linear |R| = 1.725109e-03
      1 Linear |R| = 2.889014e-15
 3 Nonlinear |R| = 2.182563e-05
      SVD: condition number 2.934736938676e+05, 0 of 301 singular values are (nearly) zero
      SVD: smallest singular values: 2.730758585276e-04 2.881279919650e-04 2.930633864015e-04 3.321877133381e-04 3.717729110401e-04
      SVD: largest singular values : 7.903903112427e+01 7.904092430221e+01 7.918950737379e+01 8.002019774275e+01 8.014058090816e+01
      0 Linear |R| = 2.182563e-05
      1 Linear |R| = 7.026828e-16
 4 Nonlinear |R| = 9.025054e-10
      SVD: condition number 2.934897633788e+05, 0 of 301 singular values are (nearly) zero
      SVD: smallest singular values: 2.730610434865e-04 2.881286828642e-04 2.930609430009e-04 3.321189912475e-04 3.717501836531e-04
      SVD: largest singular values : 7.903903177671e+01 7.904092499098e+01 7.918961651053e+01 8.002019859071e+01 8.014062104081e+01
      0 Linear |R| = 9.025054e-10
      1 Linear |R| = 3.398837e-21
 5 Nonlinear |R| = 1.847339e-14
```

With mu=10 with the old rF implementation, I get a diverged line search after 25 nonlinear iterations with SOU and convergence in 26 nonlinear iterations with QUICK ... I would say that the new implementation is quite a bit more robust!!

### CODE REVIEW

**Review by GiudGiud — DISMISSED** [review]

File: framework/include/utils/MathFVUtils.h [hunk 1]
```diff
@@ -430,10 +430,16 @@ template <typename Scalar, typename Vector>
...
-    return 1e6 * MathUtils::sign(gradC * dCD) + 0 * (gradC * dCD + phiD + phiC);
+  const auto fgrad = phiD - phiC;
+  const auto fgradC = gradC * dCD;
+  constexpr Real coeff = 1e3;
 
-  return 2. * gradC * dCD / (phiD - phiC) - 1.;
+  if (std::abs(fgradC) >= coeff * std::abs(fgrad))
+    // Second term (multiplied by zero) is to keep same sparsity pattern as else branch
+    return 2 * coeff * MathUtils::sign(fgradC) * MathUtils::sign(fgrad) +
+           0 * (gradC * dCD + phiD + phiC);
```
Comment (grmnptr) [inline framework/include/utils/MathFVUtils.h:440]: Why do we need this term?

**Review by grmnptr — COMMENTED** [review]
Just a question?

File: framework/include/utils/MathFVUtils.h (hunk 1 above)
Comment (GiudGiud) [reply framework/include/utils/MathFVUtils.h:440]: says in the comment haha

Comment (grmnptr) [reply framework/include/utils/MathFVUtils.h:440]: Ahhh, damn Peter: learn to read!
OF had a similar issue at some point, I had to add a 0.0*laplacian(variable) to make sure the offdiagonals were initialized.

Comment (grmnptr) [reply framework/include/utils/MathFVUtils.h:440]: Shouldn't vec(0)*gradC be enough? It already touches cell C and D, so we could save some operations here. Just a suggestion.

Comment (lindsayad) [reply framework/include/utils/MathFVUtils.h]: :+1: changed

Comment (grmnptr) [reply framework/include/utils/MathFVUtils.h]: Still cant change this to resolved, so feel free to resolve this of just merge if needed.

**Review by GiudGiud — DISMISSED** [review]

**Comment by lindsayad** [comment]
Ok if I use the upstream code for `rF` (e.g. not the changes here), but I bound the evaluation of psi(r) by 0 and 2, then here are the convergence histories for the lid driven problem with 10x10 mesh (recall this is the scenario in which two neighboring cells have the same value of vel-x). Minimum (but not sufficient) requirements for a scheme to be TVD is 0 <= psi(r) <= 2. Full requirement is (from Moukalled)

Screen Shot 2022-05-10 at 12 11 30 PM

### sou unbounded

```
 0 Nonlinear |R| = 6.324555e+01
      0 Linear |R| = 6.324555e+01
      1 Linear |R| = 4.170810e-14
 1 Nonlinear |R| = 3.807407e-02
      0 Linear |R| = 3.807407e-02
      1 Linear |R| = 6.772981e-14
 2 Nonlinear |R| = 6.179077e-03
      0 Linear |R| = 6.179077e-03
      1 Linear |R| = 3.785389e-13
 3 Nonlinear |R| = 5.081722e-03
      0 Linear |R| = 5.081722e-03
      1 Linear |R| = 7.788743e-13
 4 Nonlinear |R| = 4.477755e-03
      0 Linear |R| = 4.477755e-03
      1 Linear |R| = 4.752220e-15
 5 Nonlinear |R| = 4.214600e-04
      0 Linear |R| = 4.214600e-04
      1 Linear |R| = 4.483260e-16
 6 Nonlinear |R| = 6.573752e-06
      0 Linear |R| = 6.573752e-06
      1 Linear |R| = 2.313697e-18
 7 Nonlinear |R| = 2.796261e-09
      0 Linear |R| = 2.796261e-09
      1 Linear |R| = 1.254482e-21
 8 Nonlinear |R| = 1.445196e-14
```

### sou bounded

```
 0 Nonlinear |R| = 6.324555e+01
      0 Linear |R| = 6.324555e+01
      1 Linear |R| = 3.251814e-14
 1 Nonlinear |R| = 3.603600e-02
      0 Linear |R| = 3.603600e-02
      1 Linear |R| = 3.690223e-17
 2 Nonlinear |R| = 1.625687e-06
      0 Linear |R| = 1.625687e-06
      1 Linear |R| = 4.913424e-21
 3 Nonlinear |R| = 1.454895e-14
```

### quick unbounded

```
 0 Nonlinear |R| = 6.324555e+01
      0 Linear |R| = 6.324555e+01
      1 Linear |R| = 4.170810e-14
 1 Nonlinear |R| = 3.266294e-02
      0 Linear |R| = 3.266294e-02
      1 Linear |R| = 3.790502e-11
 2 Nonlinear |R| = 9.969236e-03
      0 Linear |R| = 9.969236e-03
      1 Linear |R| = 3.169328e-13
 3 Nonlinear |R| = 9.220919e-03
      0 Linear |R| = 9.220919e-03
      1 Linear |R| = 1.739155e-14
 4 Nonlinear |R| = 4.956815e-04
      0 Linear |R| = 4.956815e-04
      1 Linear |R| = 9.768500e-17
 5 Nonlinear |R| = 5.556824e-07
      0 Linear |R| = 5.556824e-07
      1 Linear |R| = 1.089497e-19
 6 Nonlinear |R| = 1.870662e-12
```

### quick bounded

```
 0 Nonlinear |R| = 6.324555e+01
      0 Linear |R| = 6.324555e+01
      1 Linear |R| = 3.251814e-14
 1 Nonlinear |R| = 3.258415e-02
      0 Linear |R| = 3.258415e-02
      1 Linear |R| = 2.905727e-17
 2 Nonlinear |R| = 1.135255e-06
      0 Linear |R| = 1.135255e-06
      1 Linear |R| = 4.709414e-21
 3 Nonlinear |R| = 1.626201e-14
```

**Comment by lindsayad** [comment]
Clearly the bounded performs much better and it's not surprising that it does so. But to me this raises the question, again, of whether it even makes sense to include SOU and QUICK as options available to the user. Why not force them to go with a scheme that is innately TVD, which precludes the comparison of phi(r) against bounds?

**Comment by GiudGiud** [comment]
so bounded SOU/QUICK performs better the proposed changes? 
Or are the results in the first post for a different case?

**Comment by lindsayad** [comment]
they were for the same case but a different PC (svd). For apples to apples here are results with changing the rF code. To directly answer your question, yes, bounding psi(r) yields better results than just changing the returned value of `Moose::FV::rF`

### rF change alone, sou

```
 0 Nonlinear |R| = 6.324555e+01
      0 Linear |R| = 6.324555e+01
      1 Linear |R| = 4.170810e-14
 1 Nonlinear |R| = 3.789924e-02
      0 Linear |R| = 3.789924e-02
      1 Linear |R| = 1.414516e-15
 2 Nonlinear |R| = 1.854462e-02
      0 Linear |R| = 1.854462e-02
      1 Linear |R| = 1.694303e-15
 3 Nonlinear |R| = 2.596531e-04
      0 Linear |R| = 2.596531e-04
      1 Linear |R| = 7.318788e-16
 4 Nonlinear |R| = 8.013175e-06
      0 Linear |R| = 8.013175e-06
      1 Linear |R| = 5.179803e-18
 5 Nonlinear |R| = 3.861665e-10
      0 Linear |R| = 3.861665e-10
      1 Linear |R| = 3.643453e-21
 6 Nonlinear |R| = 1.266413e-14
```

### rF change alone, quick

```
 0 Nonlinear |R| = 6.324555e+01
      0 Linear |R| = 6.324555e+01
      1 Linear |R| = 4.170810e-14
 1 Nonlinear |R| = 3.271746e-02
      0 Linear |R| = 3.271746e-02
      1 Linear |R| = 4.732243e-17
 2 Nonlinear |R| = 2.782304e-03
      0 Linear |R| = 2.782304e-03
      1 Linear |R| = 3.263563e-17
 3 Nonlinear |R| = 6.516954e-06
      0 Linear |R| = 6.516954e-06
      1 Linear |R| = 2.767600e-18
 4 Nonlinear |R| = 7.178573e-11
      0 Linear |R| = 7.178573e-11
      1 Linear |R| = 4.391987e-24
 5 Nonlinear |R| = 1.254839e-14
```

**Comment by lindsayad** [comment]
So what do you guys want to do? To me the options that make the most sense are:
- Remove SOU and QUICK
- Bound limiter evaluations (psi(r)) by 0 (lower) and 2 (upper) ... which is already guaranteed for all limiters except SOU and QUICK

**Comment by lindsayad** [comment]
The current state of this PR is the latter option

**Comment by lindsayad** [comment]
🎺

File: framework/include/limiters/Limiter.h [hunk 2]
```diff
@@ -64,12 +64,19 @@ class Limiter
...
+                  const VectorValue<T> * grad_phi_upwind,
+                  const RealVectorValue & dCD) const = 0;
   virtual bool constant() const = 0;
   virtual InterpMethod interpMethod() const = 0;
+  T operator()(const T & phi_upwind,
+               const T & phi_downwind,
+               const VectorValue<T> * grad_phi_upwind,
+               const RealVectorValue & dCD) const
+  {
+    return std::max(T(0), std::min(T(2), limit(phi_upwind, phi_downwind, grad_phi_upwind, dCD)));
```
Comment (grmnptr) [inline framework/include/limiters/Limiter.h:78]: What is T(0) and T(1) here? I am little confused, is T the template parameter?

**Review by grmnptr — COMMENTED** [review]
Just a few questions.

File: framework/include/limiters/Limiter.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/include/limiters/Limiter.h:78]: minmod would have T(1) not T(2) right?
I think I like this overall

[the remark this answers is not in the document]
Comment (grmnptr) [reply framework/include/limiters/Limiter.h:78]: I thinkj it would be T(0) and T(1).

[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/limiters/Limiter.h:78]: CD always returns 1. So this change has no impact on CD

File: framework/include/utils/MathFVUtils.h [hunk 3]
```diff
@@ -430,8 +430,14 @@ template <typename Scalar, typename Vector>
 Scalar
 rF(const Scalar & phiC, const Scalar & phiD, const Vector & gradC, const RealVectorValue & dCD)
 {
+  static const auto zero_vec = RealVectorValue(0);
   if ((phiD - phiC) == 0)
-    return 1e6 * MathUtils::sign(gradC * dCD) + 0 * (gradC * dCD + phiD + phiC);
+    // Handle zero denominator case. Note that MathUtils::sign returns 1 for sign(0) so we can omit
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/utils/MathFVUtils.h:435]: Yes. Constant initial condition will trigger this

File: framework/include/utils/MathFVUtils.h [hunk 4]
```diff
@@ -430,8 +430,14 @@ template <typename Scalar, typename Vector>
...
+    // that operation here (e.g. sign(phiD - phiC) = sign(0) = 1). The second term preserves the
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/utils/MathFVUtils.h:436]: > I wonder if this comparison will turn into: abs(diff) < epsilon

Is this a wondering about about (phiD - phiC) == 0 returning `true` when `phiD` and `phiC` are actually different? I expect floating point comparisons to yield more negatives than desired; not more positives. Which is exactly what I want here

[the remark this answers is not in the document]
Comment (grmnptr) [reply framework/include/utils/MathFVUtils.h:436]: Okay, so my question is the following: aside from constant initial conditions, can this be still technically (assuming floating point) 0 and not evaluate to 1? If the difference is let's say -1e-16, is this going to return `true` and `sign = 1`?

**Comment by grmnptr** [comment]
Also, I still can't mark my conversations resolved so feel free to mark them resolved!

File: framework/include/utils/MathFVUtils.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/utils/MathFVUtils.h:436]: If any of the floating point bits are different, this `if` branch will get skipped and you'll proceed to the standard evaluation of r. So if there is a 1e-16 difference (in either direction) you will skip to the main evaluation and you will get consideration of both numerator and denominator signs

[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/utils/MathFVUtils.h:436]: @grmnptr do you consider this resolved? I'm excited for your CCB elevation soon :smile:

[the remark this answers is not in the document]
Comment (grmnptr) [reply framework/include/utils/MathFVUtils.h:436]: Yes, this is resolved, thanks for the detailed answers!

**Review by GiudGiud — APPROVED** [review]

**Comment by lindsayad** [comment]
thanks guys!

**Pull request merged**